unified_planning/model/walkers/type_checker.py

```diff
@@ -83,6 +83,48 @@ def walk_param_exp(
         assert len(args) == 0
         return expression.parameter().type
 
+    def walk_always(
```

Comment:
Hi @stoppini , I unresolved these conversation because there is an error. In general, in the type_checker, you want to return None when there is a typing error or the correct type of the expression if there are no typing errors.

Here I leave you an example of the `walk_always` for clarity:
```suggestion
    def walk_always(
        self, expression: FNode, args: List["unified_planning.model.types.Type"]
    ) -> "unified_planning.model.types.Type":
        assert expression is not None
        assert expression.is_always()
        assert len(expression.args) == 1
        if args[0] is None or not args[0].is_bool_type(): #type error
            return None
        return args[0] #correct type
```

In particular, you want to check the args given as parameters for the correctness of the types, and not the `expression.args[0].type` field. Note that the `FNode.type` field is basically another call to the `type_checker`.

This change/logic needs to be applied to all the methods of the type_checker! 

Let me know if you have any doubts/questions